Rescheduled series creates invalid open meetings

oliverguenther opened the pull request:
# Ticket
https://community.openproject.org/work_packages/61729/activity

<!-- Contributors: Please check our PR guide: https://www.openproject.org/docs/development/code-review-guidelines/#preparing-your-pull-request before opening a PR. -->

<!-- Reviewers: Please check our Review guide: https://www.openproject.org/docs/development/code-review-guidelines/#reviewing -->

# What are you trying to accomplish?
Move instantiated meetings around when the schedule changes

- All n open meetings in the future will be rescheduled:
-  [x] If only the start time (hours) changed: simply update the start time hour of the instantiated meeting to match the new time
 - [x] If the date/frequency/interval, etc has changed: according ot the next n planned meetings according to the new series schedule. Even if moving from daily to monthly, or weekly to daily.
 - [x] If the user changes the # occurences so that it's fewer than actual open occurences, then you get an inline validation in the meeting details dialog.

# Merge checklist

- [x] Added/updated tests

oliverguenther committed:
Fix rescheduling of upcoming meetings

oliverguenther committed:
Add rescheduling of only start hour changed

oliverguenther committed:
Don't use filter on recurring meeting

oliverguenther committed:
Prevent changing the schedule to have less than open meetings

mrmir reviewed (changes requested):
I think something is not quite right as I get an `ActiveRecord::RecordNotUnique in RecurringMeetingsController#update` error on the recurring_meeting start_time when trying to change the schedule of the series

My steps:

- Create daily series
- Open first 3 meetings
- Change interval to 2 from 1

The tests don't count for the situation where a meeting according to the new schedule is at the same time as an open meeting with the old schedule

oliverguenther committed:
Replace unique constraint with deferrable check constraint

This allows us to change all meetings together, without hitting the
unique index

oliverguenther commented:
@mrmir I fixed that case by changing the unique index to a check constraint that we can defer to the end of the transaction.

mrmir commented:
Thanks! It works properly for the case I mentioned above, but still fails with the same error if an open meeting from the new schedule coincides with a cancelled meeting from the old schedule


Also, there's a few things still unclear to me:

1. For a meeting series with a weekly schedule, if I cancel the 2nd meeting, and then change the schedule to daily, it now shows the 8th meeting as cancelled (which is on the same day as before). I think to me it is not ideal that this happens as it isn't consistent with the moving of open meetings, where my 2nd open meeting in the old schedule becomes my 2nd open meeting in the new schedule as well, but I don't know if making the 2nd cancel meeting stay cancelled in the new schedule makes sense either. Maybe removing the cancelled meetings entirely and resetting all not 'open' meetings is better? What do you think?

2. Skipped meetings aren't considered either when the schedule is changed. If I have the 1st meeting open, 2nd skipped, and 3rd open, and change the schedule from interval 1 to interval 2, I only see the first 2 meetings open. This is again different from the behaviour of open meetings being rescheduled and also cancelled meetings being rescheduled and seems confusing. I think having it act consistently across all meeting statuses makes sense, but I don't know if that's possible for the planned occurrences

oliverguenther committed:
Remove cancelled meetings

oliverguenther commented:
> Thanks! It works properly for the case I mentioned above, but still fails with the same error if an open meeting from the new schedule coincides with a cancelled meeting from the old schedule
> 
> Also, there's a few things still unclear to me:
> 
> 1. For a meeting series with a weekly schedule, if I cancel the 2nd meeting, and then change the schedule to daily, it now shows the 8th meeting as cancelled (which is on the same day as before). I think to me it is not ideal that this happens as it isn't consistent with the moving of open meetings, where my 2nd open meeting in the old schedule becomes my 2nd open meeting in the new schedule as well, but I don't know if making the 2nd cancel meeting stay cancelled in the new schedule makes sense either. Maybe removing the cancelled meetings entirely and resetting all not 'open' meetings is better? What do you think?
> 2. Skipped meetings aren't considered either when the schedule is changed. If I have the 1st meeting open, 2nd skipped, and 3rd open, and change the schedule from interval 1 to interval 2, I only see the first 2 meetings open. This is again different from the behaviour of open meetings being rescheduled and also cancelled meetings being rescheduled and seems confusing. I think having it act consistently across all meeting statuses makes sense, but I don't know if that's possible for the planned occurrences

@mrmir  As discussed, all cancelled meetings will be removed unless only start_time_hour was changed. Added that to the expected behvior

mrmir commented:
> > Thanks! It works properly for the case I mentioned above, but still fails with the same error if an open meeting from the new schedule coincides with a cancelled meeting from the old schedule
> > Also, there's a few things still unclear to me:
> > 
> > 1. For a meeting series with a weekly schedule, if I cancel the 2nd meeting, and then change the schedule to daily, it now shows the 8th meeting as cancelled (which is on the same day as before). I think to me it is not ideal that this happens as it isn't consistent with the moving of open meetings, where my 2nd open meeting in the old schedule becomes my 2nd open meeting in the new schedule as well, but I don't know if making the 2nd cancel meeting stay cancelled in the new schedule makes sense either. Maybe removing the cancelled meetings entirely and resetting all not 'open' meetings is better? What do you think?
> > 2. Skipped meetings aren't considered either when the schedule is changed. If I have the 1st meeting open, 2nd skipped, and 3rd open, and change the schedule from interval 1 to interval 2, I only see the first 2 meetings open. This is again different from the behaviour of open meetings being rescheduled and also cancelled meetings being rescheduled and seems confusing. I think having it act consistently across all meeting statuses makes sense, but I don't know if that's possible for the planned occurrences
> 
> @mrmir As discussed, all cancelled meetings will be removed unless only start_time_hour was changed. Added that to the expected behvior

I no longer get the error when a new meeting and old cancelled meeting occur at the same time, but I also don't see cancelled meetings staying if only start_time was changed

oliverguenther committed:
Correctly deal with cancelled meetings when changing start_time_hour

oliverguenther commented:
> I no longer get the error when a new meeting and old cancelled meeting occur at the same time, but I also don't see cancelled meetings staying if only start_time was changed

@mrmir fixed and added a spec for that

mrmir commented on modules/meeting/config/locales/en.yml:
```diff
@@ -71,6 +71,13 @@ en:
+        recurring_meeting:
+          must_cover_existing_meetings:
+            one: "There is one open meeting in the series that is not covered by the new schedule. Adjust the schedule to include all existing meetings."
+            other: "There are %{count} open meetings in the series that is not covered by the new schedule. Adjust the schedule to include all existing meetings."
```
```suggestion
            other: "There are %{count} open meetings in the series that are not covered by the new schedule. Adjust the schedule to include all existing meetings."
```

mrmir reviewed (approved):
Thanks for the changes, just one string to be updated now 👍

oliverguenther committed:
Update modules/meeting/config/locales/en.yml

Co-authored-by: Mir Bhatia <m.bhatia@openproject.com>
